### PR TITLE
✨ Autoscaling: connect with rabbitmq

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR connects the autoscaling app with RabbitMQ.

- added ```GET /status``` to recover the current connection status with rabbit MQ (will be enhanced later with EC2, maybe Redis)
- tasks created through the dynamic sidecar also get the labels (user_id, study_id, uuid) NOTE: this will be at some point refactored to better names using inverse urls (see [#3638](https://github.com/ITISFoundation/osparc-simcore/issues/3638)).
- added function to ping rabbitMQ connection
- removed default values in settings RabbitSettings (this could cause some issues in the deployments - to be seen)
- prepare for simplification of tests using rabbitMQ by having a fixture that creates a rabbitmq client
- remove dependency to simcore-postgres-database package from rabbitmq_messages.py
- light refactoring

Bonus:
- add some fixes to improve reliability of test_rabbit.py (webserver integration 02), fixes https://github.com/ITISFoundation/osparc-simcore/issues/3574
- add some fixes in python-socketio server to correctly cancel its tasks, as discussed here: https://github.com/miguelgrinberg/python-socketio/discussions/1092
closing the python-socketio AsyncServer needs some ad-hoc removal.

## Related issue/s

needed for https://github.com/ITISFoundation/osparc-simcore/issues/3341

### CODE REVIEW

File: packages/models-library/src/models_library/docker.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import re
-from typing import Optional
+from typing import Final, Optional
 
 from pydantic import ConstrainedStr, constr
 
+_DOCKER_LABEL_KEY_REGEX: Final[re.Pattern] = re.compile(
+    r"^(?!(\.|\-|com.docker\.|io.docker\.|org.dockerproject\.))(?!.*(--|\.\.))[a-z0-9\.-]+$"
```
Comment (pcrespov) [inline packages/models-library/src/models_library/docker.py:7]: can you please add a test to check this PR? 
It is very handy to have one at hand in case we find a corner case 
You can add one very fast using ``packages/models-library/tests/test_basic_regex.py``

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py [hunk 2]
```diff
@@ -82,22 +91,19 @@ async def check_dynamic_resources(app: FastAPI) -> None:
                     InstanceTypeType, ec2_instances_needed[0].name
                 ),
                 tags={
-                    "io.osparc.autoscaling.created": f"{datetime.utcnow()}",
-                    "io.osparc.autoscaling.version": f"{VERSION}",
-                    "io.osparc.autoscaling.monitored_nodes_labels": json.dumps(
+                    "io.simcore.autoscaling.created": f"{datetime.utcnow()}",
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:94]: Q? why ``io.simcore`` and not ``io.osparc``. The convention is to use the reversed nmap i.e. the latter

File: services/autoscaling/src/simcore_service_autoscaling/models.py [hunk 3]
```diff
@@ -10,3 +14,28 @@ class EC2Instance(BaseModel):
     name: str
     cpus: PositiveInt
     ram: ByteSize
+
+
+class SimcoreServiceDockerLabelKeys(BaseModel):
+    # NOTE: in a next PR, this should be moved to packages models-library and used
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/models.py:20]: i wonder how many variants we have of this :)

File: services/autoscaling/src/simcore_service_autoscaling/rabbitmq.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from servicelib.rabbitmq import RabbitMQClient
+from servicelib.rabbitmq_utils import wait_till_rabbitmq_responsive
+from settings_library.rabbit import RabbitSettings
+
+from .core.errors import ConfigurationError
+
+logger = logging.getLogger(__name__)
+
+
+def setup(app: FastAPI) -> None:
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/rabbitmq.py:17]: very clean an clear!

File: services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/dask_scheduler.py [hunk 5]
```diff
@@ -222,7 +222,7 @@ async def _task_state_change_handler(self, event: str) -> None:
                 project_id=project_id,
                 node_id=node_id,
                 service_uuid=node_id,
-                service_type=NodeClass.COMPUTATIONAL,
+                service_type=f"{NodeClass.COMPUTATIONAL}",
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/comp_scheduler/dask_scheduler.py:225]: TIP: ``NodeClass.COMPUTATIONAL.value``?

**Review by pcrespov — APPROVED** [review]
Nice!

File: packages/models-library/src/models_library/docker.py (hunk 1 above)
Comment (sanderegg) [reply packages/models-library/src/models_library/docker.py:7]: of course

**Review by GitHK — APPROVED** [review]
Very nice: one question below about rabbit being not properly configured. I'm wondering what happens if there was a mistake in the DNS, when and where would that be seen

File: services/autoscaling/src/simcore_service_autoscaling/models.py (hunk 3 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/models.py:20]: quite a few... quite  a few..

File: services/autoscaling/src/simcore_service_autoscaling/rabbitmq.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+def get_rabbitmq_client(app: FastAPI) -> RabbitMQClient:
+    if not app.state.rabbitmq_client:
+        raise ConfigurationError(
+            msg="RabbitMQ client is not available. Please check the configuration."
+        )
+    return cast(RabbitMQClient, app.state.rabbitmq_client)
+
+
+async def post_message(app: FastAPI, message: RabbitMessageBase) -> None:
+    with log_catch(logger, reraise=False), contextlib.suppress(ConfigurationError):
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/rabbitmq.py:46]: yes, it does check on startup.
- if the RABBIT entries are not set, there will just be a warning saying it is disabled as this is ok
- if the RABBIT entries are set, but wrong, then the autoscaling app will not start and complain

File: services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py (hunk 2 above)
Comment (pcrespov) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:94]: According to the key format recommendations https://docs.docker.com/config/labels-custom-metadata/#key-format-recommendations ... it should be the reverse DNS of a domain that **we own**, therefore I would say it should be ``io.osparc``.

image


 In any case, we have to refactor this also in other parts.  I saw you created a case for that

Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py:94]: I agree with you. but still since these labels are not public there is no harm done.
